### PR TITLE
Introduce upgrade --from-manifests flag

### LINKED ISSUE #2629: RFC: load upgrade configuration from kubernetes manifest
### What problem are you trying to solve?

We use the gitops method for maintaining our clusters and deploying services. This involves writing kubernetes manifests (or kustomize, or helm charts) to a config repo and ArgoCD applies them to the cluster specified.

For example:

```text
infrastructure
  \__ linkerd
         \__ azure
                \__ dev-ea
                \__ qa-scu
                \__ prod-scu
```

each folder under the ```azure``` folder is a specific cluster and contains the output of `linkerd install --ha --proxy-auto-inject > install.yaml` as well as an ingress for the web-ui, etc.

With the new upgrade mechanism (and identity), we can no longer do a `linkerd install --ha --proxy-auto-inject > install.yaml` (even with `--ignore-cluster`) and expect identity to stay happy (due to install creating new keys, etc).

If we do a `linkerd upgrade > install.yaml` it will take the config from whichever cluster is my current kubernetes context. Therefore if I wish to upgrade ```prod-scu``` and I'm not paying attention, it may write config values from ```dev-ea```.

### How should the problem be solved?

The addition of a `--from-manifest` flag for the `linkerd upgrade` command. Instead of reading the config from the `linkerd-config` configmap, it reads it from the manifest created by a prior `linkerd install --flags... > install.yaml`

### Any alternatives you've considered?

Nothing else on the linkerd side comes to mind. We could probably write some tooling that does the upgrade in a way that mitigates the risk outlined above.

### How would users interact with this feature?

```linkerd upgrade --from-manifests install.yaml > install.yaml```

### PR DESCRIPTION
**Pull request by siggy**
The `linkerd upgrade` command read the control-plane's config from
Kubernetes, which required the environment to be configured to connect
to the appropriate k8s cluster.

Introduce a `linkerd upgrade --from-manifests` flag, allowing the user
to feed the output of `linkerd install` into the upgrade command.

Fixes #2629

Signed-off-by: Andrew Seigner <siggy@buoyant.io>

Relates to #2637.

## Notes

Note that the first commit, 5fb1c0c, contains all the code changes. The second commit simply renames `test_helper.go` to `fake.go`.

### Install flow

```bash
linkerd install > install_manifests.yaml
kubectl apply -f install_manifests.yaml
linkerd upgrade --from-manifests install_manifests.yaml
```

### Validate `linkerd upgrade` command works the same with and without `--from-manifests`

```bash
linkerd install | kubectl apply -f -

diff <(
  bin/linkerd upgrade
) <(
  kubectl -n linkerd get cm/linkerd-config secrets/linkerd-identity-issuer -oyaml |
    bin/linkerd upgrade --from-manifests -
)
```

### CODE REVIEW

**Comment by siggy** [comment]
@jon-walton if you have a moment, please take this out for a spin, thanks!

**Comment by jon-walton** [comment]
awesome, that works :+1: thanks! 🎉

File: cli/cmd/upgrade.go [hunk 1]
```diff
@@ -22,12 +22,16 @@ const (
...
-)
+type upgradeOptions struct {
+	manifests string
+	*installOptions
+}
 
 func newUpgradeOptionsWithDefaults() *upgradeOptions {
-	return &upgradeOptions{newInstallOptionsWithDefaults()}
+	return &upgradeOptions{
+		"",
```
Comment (olix0r) [inline cli/cmd/upgrade.go]: tioli: named params might be easier to read here

**Comment by grampelberg** [comment]
@olix0r because they show up in the manifest?

**Comment by ihcsim** [comment]
Just wanna confirm that we are ok with writing the issuer's secret to disk.

**Comment by siggy** [comment]
@ihcsim Re: writing secrets to disk, I think that's a good question for @grampelberg, and if the answer is "yes", maybe deserves treatment in our docs.

**Comment by olix0r** [comment]
@grampelberg right. Ideally, we'd encrypt that part of the manifest and have `--from-manifest` decrypt the secrets

**Comment by ihcsim** [comment]
fwiw, I think all #2629 cares about is the `linkerd-config` config map.

**Comment by alpeb** [comment]
Another possibility would be to add a note in the Upgrade CLI docs, telling people they can store all the resources in disk, but discouraging keeping the Secret one. Then we could have `fetchIdentityValues()` not fail when the Secret is not found when using `--from-manifests`. In my brief testing this doesn't break anything.
One downside I can see is when trying to do `kubectl apply --prune` with resource files in disk that don't match what there is in the cluster, one has to be careful not to delete the Secret in the cluster.

**Comment by jon-walton** [comment]
This is an interesting point.

However, when using a gitops workflow, a `linkerd install --flags > install.yaml` would end up in git and deployed by the deployment controller (in my case, ArgoCD). if the secret is kept out of the manifest, we'll need linkerd to create the secret and store it in a k8s secret object at runtime

**Comment by grampelberg** [comment]
Is there another option to writing the secret to disk? It'd need to live somewhere ...

From a docs perspective, it is pretty easy to suggest piping through `gpg` or the like before committing the file.

**Comment by ihcsim** [comment]
> if the secret is kept out of the manifest, we'll need linkerd to create the secret and store it in a k8s secret object at runtime

The (identity) secret is created during installation, and remained untouched, in the cluster, during an upgrade.

**Comment by siggy** [comment]
Going to merge this as-is. I agree storing secrets in the clear on disk is not great, but also agree we can guide users via docs on best practices, such as @grampelberg's `gpg` approach.

**Comment by grampelberg** [comment]
Can we open an issue on the website repo to track writing up a blurb about managing the manifest?

**Comment by siggy** [comment]
@grampelberg filed linkerd/website#275

**Pull request merged**